cmd/query/app/querysvc/adjuster/spaniduniquifier.go

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2019 The Jaeger Authors.
+// Copyright (c) 2017 Uber Technologies, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package adjuster
+
+import (
+	"errors"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+
+	"github.com/jaegertracing/jaeger/internal/jptrace"
+)
+
+var _ Adjuster = (*SpanIDDeduper)(nil)
+
+const (
+	warningTooManySpans = "cannot assign unique span ID, too many spans in the trace"
+)
+
+// SpanIDUniquifier returns an adjuster that changes span ids for server
+// spans (i.e. spans with tag: span.kind == server) if there is another
+// client span that shares the same span ID. This is needed to deal with
+// Zipkin-style clients that reuse the same span ID for both client and server
+// side of an RPC call. Jaeger UI expects all spans to have unique IDs.
+//
+// This adjuster never returns any errors. Instead it records any issues
+// it encounters in Span.Warnings.
+func SpanIDUniquifier() *SpanIDDeduper {
+	return &SpanIDDeduper{}
+}
+
+type SpanIDDeduper struct {
+	maxUsedID pcommon.SpanID
```

Comment:
cannot have state in the object returned from the constructor, because the object is supposed to be reused / concurrent. This is where Func might be useful:
```
func SpanIDUniquifier() Adjuster {
  return Func(func(traces ptrace.Traces) error {
    adjuster := new(spanIDUniquifier) // type can be private
    return adjuster.adjust(traces) // method can be private
  })
}
```

Reply:
done